### PR TITLE
8320522: Remove code related to `RegisterFinalizersAtInit`

### PR DESCRIPTION
**Pull request by DanHeidinga**
Remove the code related to -XX:[+-]RegisterFinalizersAtInit in JDK23.

`make test-tier1` passed with this change

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320522](https://bugs.openjdk.org/browse/JDK-8320522): Remove code related to `RegisterFinalizersAtInit` (**Enhancement** - P4)

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [eb0ba81f](https://git.openjdk.org/jdk/pull/18823/files/eb0ba81fde975cbf9d2216c214c81447b0f0ddc9)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18823/head:pull/18823` \
`$ git checkout pull/18823`

Update a local copy of the PR: \
`$ git checkout pull/18823` \
`$ git pull https://git.openjdk.org/jdk.git pull/18823/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18823`

View PR using the GUI difftool: \
`$ git pr show -t 18823`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18823.diff">https://git.openjdk.org/jdk/pull/18823.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18823#issuecomment-2061983250)

### CODE REVIEW

File: src/hotspot/share/classfile/classFileParser.cpp [hunk 1]
```diff
@@ -4188,8 +4188,7 @@ void ClassFileParser::set_precomputed_flags(InstanceKlass* ik) {
   // If it cannot be fast-path allocated, set a bit in the layout helper.
   // See documentation of InstanceKlass::can_be_fastpath_allocated().
   assert(ik->size_helper() > 0, "layout_helper is initialized");
-  if ((!RegisterFinalizersAtInit && ik->has_finalizer())
-      || ik->is_abstract() || ik->is_interface()
+  if (ik->is_abstract() || ik->is_interface()
```
Comment (coleenp) [inline src/hotspot/share/classfile/classFileParser.cpp:4191]: TIL what this bit was.  There are some comments in the interpreters that claim that the bit will be set if there are finalizers, which is no longer true.  The comments look like:

  // test to see if it has a finalizer or is malformed in some way

Comment (DanHeidinga) [reply src/hotspot/share/classfile/classFileParser.cpp:4191]: Thanks.  I never would have found that!

**Review by coleenp — APPROVED** [review]
Looks good!  abstractInterpreter.cpp needs a copyright update.

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 2]
```diff
@@ -1511,14 +1511,7 @@ instanceOop InstanceKlass::register_finalizer(instanceOop i, TRAPS) {
 instanceOop InstanceKlass::allocate_instance(TRAPS) {
   bool has_finalizer_flag = has_finalizer(); // Query before possible GC
```
[the remark this answers is not in the document]
Comment (DanHeidinga) [reply src/hotspot/share/oops/instanceKlass.cpp]: Thanks.  Pushed an update

**Review by albertnetymk — APPROVED** [review]

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Pull request closed without merging**